### PR TITLE
feat(amf): adding support for registration reject

### PR DESCRIPTION
**Pull request by Chandhu-Wavelabs**
## Summary

Registration reject support added

## Test Plan

unit test completed

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by panyogesh** [comment]
Related to ZenHub Task : #6540

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp [hunk 1]
```diff
@@ -259,6 +261,27 @@ TEST(test_smf_context_struct, test_smf_context_creation) {
...
+  reg_rej.message_type.msg_type                                        = 0x44;
+  reg_rej.m5gmm_cause.m5gmm_cause                                      = 23;
+
+  bool encode_res = false;
+
+  uint32_t len = 4;
+
+  encode_res = encode_registration_reject_msg(
+      &reg_rej, nas5g_pkt_snap.registration_reject, len);
+
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf.cpp]: can you also decode the encoded buffer and assert that the decoded messages matches with reg_rej?

**Comment by ulaskozat** [comment]
There are bunch of memory alignment warnings. Could you check them as they can be quite serious?

File: lte/gateway/c/core/oai/test/amf/test_amf.cpp (hunk 1 above)
Comment (Chandhu-Wavelabs) [reply lte/gateway/c/core/oai/test/amf/test_amf.cpp]: added support for decoding encoded buffer and matching with reg_rej.

**Review by ulaskozat — APPROVED** [review]
LGTM

File: lte/gateway/c/core/oai/tasks/amf/Registration.cpp [hunk 2]
```diff
@@ -302,6 +307,7 @@ static int amf_registration_reject(
     amf_data_sec.amf_as_set_security_data(
         &amf_sap.u.amf_as.u.establish.sctx, NULL, false, false);
   }
+  OAILOG_INFO(LOG_NAS_AMF, "Processing REGITRATION_REJECT message\n");
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/amf/Registration.cpp]: change to DEBUG level. REGITRATION_REJECT -> REGISTRATION_REJECT.

Comment (Kaleem-Wavelabs) [reply lte/gateway/c/core/oai/tasks/amf/Registration.cpp]: Done

**Comment by Kaleem-Wavelabs** [comment]
> There are bunch of memory alignment warnings. Could you check them as they can be quite serious?
@ulaskozat 
1) For the below warning, observed that the passing pointer arguments such as registration ies, common/specific procedures were all constructed using new operator on structure objects. There were no constructors defined.
Here delete is used to match the new operations used previously to allocate the memory.
So delete on these pointer arguments would only deallocate the memory without need to call destructors, as there were not destructors defined. 

_Check warning on line 1096 in lte/gateway/c/core/oai/tasks/amf/Registration.cpp

GitHub Actions
/ build_oai
lte/gateway/c/core/oai/tasks/amf/Registration.cpp#L1096
deleting 'void*' is undefined [-Wdelete-incomplete]_

2) For the below set of warning(s), fixed the structure definition without "__attribute__((__packed__))" to be inlign with to refer the members with the pointers.

 Check warning on line 245 in lte/gateway/c/core/oai/tasks/amf/amf_as.cpp

GitHub Actions
/ build_oai
lte/gateway/c/core/oai/tasks/amf/amf_as.cpp#L245
taking address of packed member of 'amf_as_message_s::<unnamed union>' may result in an unaligned pointer value [-Waddress-of-packed-member]

**Comment by ulaskozat** [comment]
@rdefosse can you check why OAI Jenkins pipeline is failing?

**Review by ulaskozat — CHANGES_REQUESTED** [review]
@Chandhu-Wavelabs there is another legit warning on delete_wrapper. you should use free not delete.

**Comment by Kaleem-Wavelabs** [comment]
> @Chandhu-Wavelabs there is another legit warning on delete_wrapper. you should use free not delete.

@ulaskozat 
Observed that the passing pointer arguments such as registration ies, common/specific procedures were all constructed using new operator on structure objects. There were no constructors defined.
Here delete is used to match the new operations used previously to allocate the memory.
So delete on these pointer arguments would only deallocate the memory without need to call destructors, as there were not destructors defined.
Please let us know your suggestion.

**Comment by ulaskozat** [comment]
> > @Chandhu-Wavelabs there is another legit warning on delete_wrapper. you should use free not delete.
> 
> @ulaskozat
> Observed that the passing pointer arguments such as registration ies, common/specific procedures were all constructed using new operator on structure objects. There were no constructors defined.
> Here delete is used to match the new operations used previously to allocate the memory.
> So delete on these pointer arguments would only deallocate the memory without need to call destructors, as there were not destructors defined.
> Please let us know your suggestion.

then please define the deconstructors and do not delete "void". So either follow the pattern of malloc/free with (void *) or use new/delete with proper data structures.

**Comment by rdefosse** [comment]
> 
> 
> @rdefosse can you check why OAI Jenkins pipeline is failing?

@ulaskozat 

I'm just coming back from vacation. Something went wrong during deployment.

When the contributor fixes the requested changes, let see if it occurs again.

Raphael

**Comment by Kaleem-Wavelabs** [comment]
> > > @Chandhu-Wavelabs there is another legit warning on delete_wrapper. you should use free not delete.
> > 
> > 
> > @ulaskozat
> > Observed that the passing pointer arguments such as registration ies, common/specific procedures were all constructed using new operator on structure objects. There were no constructors defined.
> > Here delete is used to match the new operations used previously to allocate the memory.
> > So delete on these pointer arguments would only deallocate the memory without need to call destructors, as there were not destructors defined.
> > Please let us know your suggestion.
> 
> then please define the deconstructors and do not delete "void". So either follow the pattern of malloc/free with (void *) or use new/delete with proper data structures.

we are implementing the changes as suggested.

**Comment by ulaskozat** [comment]
rebase and push.

**Comment by Chandhu-Wavelabs** [comment]
> rebase and push.

Done

**Review by ulaskozat — APPROVED** [review]
LGTM

**Comment by ulaskozat** [comment]
Since we are tracking "delete_wrapper" as a separate issue and it is not part of this PR, I merged this PR.

**Pull request merged**